addons/bus/tests/common.py

```diff
@@ -0,0 +1,95 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import struct
+from threading import Event
+import unittest
+from unittest.mock import patch
+
+try:
+    import websocket
+except ImportError:
+    websocket = None
+
+import odoo.tools
+from odoo.tests import HOST, common
+from ..websocket import CloseCode, WebsocketConnectionHandler
+
+
+class WebsocketCase(common.HttpCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        if websocket is None:
+            cls._logger.warning("websocket-client module is not installed")
+            raise unittest.SkipTest("websocket-client module is not installed")
+        cls._WEBSOCKET_URL = f"ws://{HOST}:{odoo.tools.config['http_port']}/websocket"
+
+    def setUp(self):
+        super().setUp()
+        self._websockets = set()
+        # Used to ensure websocket connections have been closed
+        # properly.
+        self._websocket_events = set()
+        original_serve_forever = WebsocketConnectionHandler._serve_forever
+
+        def _mocked_serve_forever(*args):
+            websocket_closed_event = Event()
+            self._websocket_events.add(websocket_closed_event)
+            original_serve_forever(*args)
+            websocket_closed_event.set()
+
+        self._serve_forever_patch = patch.object(
+            WebsocketConnectionHandler,
+            '_serve_forever',
+            wraps=_mocked_serve_forever
+        )
+        self._serve_forever_patch.start()
+
+    def tearDown(self):
+        # ensure websockets are closed after each test.
+        self._close_websockets()
+        self._serve_forever_patch.stop()
+        super().tearDown()
```

Comment:
I agree for `serve_forever_patch` but not for `close_websockets`. First, if `setUp` fails, no websocket will be opened and so there will be nothing to close. More important, the `close_websocket` method must be called before `tearDown` to prevent issues with test cursors and rollbacks (it waits for `server_forever` to complete, ensuring the cursor opened by the lifecycle events is properly closed).